File: io/assets/AssetLoader.java
package common.io.assets;

import common.CommonStatic;
import common.io.PackLoader;
import common.io.PackLoader.Preload;
import common.io.PackLoader.ZipDesc;
import common.io.PackLoader.ZipDesc.FileDesc;
import common.io.assets.Admin.StaticPermitted;
import common.io.assets.AssetLoader.AssetHeader.AssetEntry;
import common.io.json.JsonClass;
import common.io.json.JsonField;
import common.pack.Context;
import common.pack.Context.ErrType;
import common.pack.PackData.PackDesc;
import common.system.files.FDFile;
import common.system.files.VFile;
import common.system.files.VFileRoot;
import common.util.Data;

import java.io.*;
import java.util.*;
import java.util.Map.Entry;
import java.util.function.Consumer;

@StaticPermitted
public class AssetLoader {

	@JsonClass
	public static class AssetHeader {

		@JsonClass
		public static class AssetEntry {

			@JsonField
			public final PackDesc desc;

			@JsonField
			public final int size;

			public AssetEntry() {
				desc = null;
				size = 0;
			}

			private AssetEntry(File f) throws Exception {
				desc = PackLoader.readPack((fd) -> false, f).desc;
				size = (int) f.length();
			}

			@Override
			public boolean equals(Object o) {
				return (o instanceof AssetEntry) && ((AssetEntry) o).desc.id.equals(desc.id);
			}

		}

		@JsonField(generic = AssetEntry.class)
		public final ArrayList<AssetEntry> list = new ArrayList<>();

		public int offset;

		public AssetHeader() {
		}

		private boolean add(File file) throws Exception {
			AssetEntry ent = new AssetEntry(file);
			if (list.contains(ent))
				return false;
			list.add(ent);
			return true;
		}

	}

	public static final String CORE_VER = "0.7.4.3";

	private static final String[] ANIMFL = { ".imgcut", ".mamodel", ".maanim" };
	private static final String[] NONPRE = { "\\./org/img/../.....\\.png", "\\./org/enemy/.../..._.\\.png",
			"\\./org/unit/..././..._.\\.png", "\\./org/unit/..././udi..._.\\.png" };

	private static final int LEN = 1024;

	public static void load(Consumer<Double> prog) {
		try {
			File folder = CommonStatic.ctx.getAssetFile("./assets/");
			TreeMap<String, File> map = new TreeMap<>();
			for (File f : folder.listFiles())
				if (f.getName().endsWith(".assets.bcuzips"))
					map.put(f.getName(), f);
			int i = 0;
			TreeMap<String, ZipDesc> zips = new TreeMap<>();
			for (File f : map.values()) {
				int I = i;
				prog.accept(1.0 * (i++) / map.size());
				Consumer<Double> con = (d) -> prog.accept((I + d) / map.size());
				List<ZipDesc> list = PackLoader.readAssets(AssetLoader::getPreload, f, con);
				for (ZipDesc zip : list)
					if (Data.getVer(zip.desc.BCU_VERSION) <= Data.getVer(CORE_VER))
						zips.put(zip.desc.id, zip);
			}
			for (ZipDesc zip : zips.values())
				VFile.getBCFileTree().merge(zip.tree);
		} catch (Exception e) {
			CommonStatic.ctx.noticeErr(e, ErrType.FATAL, "failed to read asset");
		}
		try {
			File folder = new File(CommonStatic.ctx.getBCUFolder(), "./assets/custom");
			if (folder.exists()) {
				VFileRoot vr = new VFileRoot(".");
				add(vr, folder);
				VFile.getBCFileTree().merge(vr);
			}
		} catch (Exception e) {
			CommonStatic.ctx.noticeErr(e, ErrType.FATAL, "failed to read custom asset");
		}
	}

	public static void merge() throws Exception {
		try {
			File folder = CommonStatic.ctx.getAssetFile("./assets/");
			Map<String, Map<String, File>> map = new TreeMap<>();
			for (File f : folder.listFiles()) {
				if(f.getName().endsWith("custom.asset.bcuzip")) {
					Map<String, File> sub = map.computeIfAbsent("custom", k -> new TreeMap<>());

					sub.put("custom", f);

					continue;
				}

				if (f.getName().endsWith(".asset.bcuzip")) {
					String pre = f.getName().substring(0, 2);
					String name = f.getName().substring(0, 6);
					Map<String, File> sub = map.computeIfAbsent(pre, k -> new TreeMap<>());
					sub.put(name, f);
				}
			}

			for (Entry<String, Map<String, File>> emain : map.entrySet()) {
				String targetName;

				if(emain.getKey().equals("custom")) {
					targetName = "./assets/"+emain.getKey()+".assets.bcuzips";
				} else {
					targetName = "./assets/" + emain.getKey() + "xxxx.assets.bcuzips";
				}

				File target = CommonStatic.ctx.getAssetFile(targetName);
				File dst = CommonStatic.ctx.getAssetFile("./assets/.assets.bcuzips.temp");
				Context.check(dst);
				FileOutputStream fos = new FileOutputStream(dst);
				AssetHeader header = new AssetHeader();
				FileInputStream fis = null;
				if (target.exists()) {
					fis = new FileInputStream(target);
					PackLoader.read(fis, header);
				}
				List<File> queue = new ArrayList<>();
				for (Entry<String, File> esub : emain.getValue().entrySet())
					if (header.add(esub.getValue()))
						queue.add(esub.getValue());
					else
						Context.delete(esub.getValue());
				PackLoader.write(fos, header);
				if (fis != null) {
					stream(fos, fis);
					fis.close();
				}
				for (File efile : queue) {
					fis = new FileInputStream(efile);
					stream(fos, fis);
					fis.close();
					if(!efile.getName().endsWith("custom.asset.bcuzip"))
						Context.delete(efile);
				}
				fos.close();
				Context.delete(target);
				dst.renameTo(target);
			}
		} catch (Exception e) {
			CommonStatic.ctx.noticeErr(e, ErrType.FATAL, "failed to merge asset");
		}
	}

	public static Set<String> previewAssets() {
		try {
			File folder = CommonStatic.ctx.getAssetFile("./assets/");
			Set<String> ans = new TreeSet<>();
			if (!folder.exists())
				return ans;
			for (File f : folder.listFiles()) {
				if (f.getName().endsWith(".assets.bcuzips")) {
					AssetHeader header = new AssetHeader();
					FileInputStream fis = new FileInputStream(f);
					PackLoader.read(fis, header);
					fis.close();
					for (AssetEntry ent : header.list)
						ans.add(ent.desc.id);
				}
				if (f.getName().endsWith(".asset.bcuzip")) {
					ans.add(PackLoader.readPack((fd) -> false, f).desc.id);
				}
			}
			return ans;
		} catch (Exception e) {
			CommonStatic.ctx.noticeErr(e, ErrType.FATAL, "failed to preview asset");
			return null;
		}
	}

	private static void add(VFile vf, File f) {
		for (File fi : f.listFiles())
			if (fi.isDirectory())
				add(new VFile(vf, fi.getName()), fi);
			else
				new VFile(vf, fi.getName(), new FDFile(fi));
	}

	private static Preload getPreload(ZipDesc desc) {
		return Data.getVer(desc.desc.BCU_VERSION) > Data.getVer(CORE_VER) ? (fd) -> false : AssetLoader::preload;
	}

	private static boolean preload(FileDesc fd) {
		if (fd.size < 1024)
			return true;
		for (String str : ANIMFL)
			if (fd.path.endsWith(str))
				return false;
		for (String str : NONPRE)
			if (fd.path.length() == str.length() - 2 && fd.path.matches(str))
				return false;
		return true;
	}

	private static void stream(OutputStream fos, InputStream fis) throws IOException {
		byte[] data = new byte[LEN];
		int len;
		do {
			len = fis.read(data);
			fos.write(data, 0, len);
		} while (len == LEN);
	}

}

File: io/assets/AssetLoader.kt
package common.io.assets

import common.CommonStatic
import common.io.PackLoader
import common.io.PackLoader.Preload
import common.io.PackLoader.ZipDesc
import common.io.PackLoader.ZipDesc.FileDesc
import common.io.json.JsonClass
import common.io.json.JsonField
import common.pack.Context
import common.pack.Context.ErrType
import common.pack.PackData.PackDesc
import common.system.files.VFile
import common.util.Data
import java.io.*
import java.io.Fileimport
import java.util.*

com.google.api.client.json.jackson2.JacksonFactory
import kotlin.Throws
import java.io.IOException
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.TreeMap
import java.util.TreeSet

object AssetLoader {
    const val CORE_VER = "0.5.0.0"
    private val ANIMFL = arrayOf(".imgcut", ".mamodel", ".maanim")
    private val NONPRE = arrayOf("\\./org/img/../.....\\.png", "\\./org/enemy/.../..._.\\.png",
            "\\./org/unit/..././..._.\\.png", "\\./org/unit/..././udi..._.\\.png")
    private const val LEN = 1024
    fun load() {
        try {
            val folder: File = CommonStatic.ctx.getAssetFile("./assets/")
            for (f in folder.listFiles()) {
                if (f.name.endsWith(".assets.bcuzips")) {
                    val list: List<ZipDesc> = PackLoader.readAssets(PackLoader.Preloader { obj: ZipDesc? -> getPreload() }, f)
                    for (zip in list) if (Data.Companion.getVer(zip.desc.BCU_VERSION) <= Data.Companion.getVer(CORE_VER)) {
                        VFile.Companion.getBCFileTree().merge(zip.tree)
                    }
                }
            }
        } catch (e: Exception) {
            CommonStatic.ctx.noticeErr(e, ErrType.FATAL, "failed to read asset")
        }
    }

    @Throws(Exception::class)
    fun merge() {
        try {
            val folder: File = CommonStatic.ctx.getAssetFile("./assets/")
            val map: MutableMap<String, MutableMap<String, File>> = TreeMap<String, Map<String, File>>()
            for (f in folder.listFiles()) {
                if (f.name.endsWith(".asset.bcuzip")) {
                    val pre = f.name.substring(0, 2)
                    val name = f.name.substring(0, 6)
                    var sub = map[pre]
                    if (sub == null) map[pre] = TreeMap<String, File>().also { sub = it }
                    sub!![name] = f
                }
            }
            for ((key, value) in map) {
                val target: File = CommonStatic.ctx.getAssetFile("./assets/" + key + "xxxx.assets.bcuzips")
                val dst: File = CommonStatic.ctx.getAssetFile("./assets/.temp.assets.bcuzips")
                Context.Companion.check(dst)
                val fos = FileOutputStream(dst)
                val header = AssetHeader()
                var fis: FileInputStream? = null
                if (target.exists()) {
                    fis = FileInputStream(target)
                    PackLoader.read(fis, header)
                }
                val queue: MutableList<File> = ArrayList()
                for ((_, value1) in value) if (header.add(value1)) queue.add(value1) else Context.Companion.delete(value1)
                PackLoader.write(fos, header)
                if (fis != null) {
                    stream(fos, fis)
                    fis.close()
                }
                for (efile in queue) {
                    fis = FileInputStream(efile)
                    stream(fos, fis)
                    fis.close()
                    Context.Companion.delete(efile)
                }
                fos.close()
                Context.Companion.delete(target)
                dst.renameTo(target)
            }
        } catch (e: Exception) {
            // TODO corruption prevention
            CommonStatic.ctx.noticeErr(e, ErrType.FATAL, "failed to merge asset")
        }
    }

    fun previewAssets(): Set<String>? {
        return try {
            val folder: File = CommonStatic.ctx.getAssetFile("./assets/")
            val ans: MutableSet<String> = TreeSet<String>()
            for (f in folder.listFiles()) {
                if (f.name.endsWith(".assets.bcuzips")) {
                    val header = AssetHeader()
                    val fis = FileInputStream(f)
                    PackLoader.read(fis, header)
                    fis.close()
                    for (ent in header.list) ans.add(ent.desc.id)
                }
            }
            ans
        } catch (e: Exception) {
            CommonStatic.ctx.noticeErr(e, ErrType.FATAL, "failed to preview asset")
            null
        }
    }

    private fun getPreload(desc: ZipDesc): Preload {
        return if (Data.Companion.getVer(desc.desc.BCU_VERSION) > Data.Companion.getVer(CORE_VER)) Preload { fd: FileDesc? -> false } else Preload { obj: FileDesc? -> preload() }
    }

    private fun preload(fd: FileDesc): Boolean {
        if (fd.size < 1024) return true
        for (str in ANIMFL) if (fd.path.endsWith(str)) return false
        for (str in NONPRE) if (fd.path.length == str.length - 2 && fd.path.matches(str)) return false
        return true
    }

    @Throws(IOException::class)
    private fun stream(fos: OutputStream, fis: InputStream?) {
        val data = ByteArray(LEN)
        var len: Int
        do {
            len = fis!!.read(data)
            fos.write(data, 0, len)
        } while (len == LEN)
    }

    @JsonClass
    class AssetHeader {
        @JsonClass
        class AssetEntry {
            @JsonField
            val desc: PackDesc?

            @JsonField
            val size: Int

            constructor() {
                desc = null
                size = 0
            }

            constructor(f: File) {
                desc = PackLoader.readPack(Preload { fd: FileDesc? -> false }, f).desc
                size = f.length().toInt()
            }

            override fun equals(o: Any?): Boolean {
                return o is AssetEntry && o.desc.id == desc.id
            }
        }

        @JsonField(generic = [AssetEntry::class])
        val list = ArrayList<AssetEntry>()
        var offset = 0

        @Throws(Exception::class)
        fun add(file: File): Boolean {
            val ent = AssetEntry(file)
            if (list.contains(ent)) return false
            list.add(ent)
            return true
        }
    }
}
